rfd/0058-package-distribution.md

```diff
@@ -0,0 +1,66 @@
+
+Fixing the current solution would allow for channel support via standard practices for APT using APT distribution and component parameters. This is inline with what most APT repositories do, and supports the scheme that teleport users are most familiar with.
+
+#### YUM
+As with APT, the current solutoin does not support YUM channels. GCP would add support, but only by means of the project ID and repository name. As with APT this is a somewhat fragile and inflexible solution. Fixing the current solution would allow for channel support in whatever manner we like, allowing it to conform to best practices and standard naming scheme.
+
+### Recommendations
+It is worth noting that GCP Artifact Registry is in preview and has not had an update since 11/2021. It's lacking some features as outlined above as well as some other we may wish to use in the future (i.e. bringing our own key). In addition, if there was a disaster recovery event where we needed to push to an Artifact Registry in another region while GCP had an outage, we would break `apt update` and `yum update` for our customers as the key they would use is region-specific.
+
+As discussed above, JFrog Artifactory and PackageCloud are non-starters due to their signing key requirements.
+
+As a consequence of the the issues with the third party solutions, along with the additional channel support features available, I recommend fixing our current S3-hosted solution. This would require switching reprepo for aptly in our Drone pipelines, and updating the current createrepo steps.
+
+The follwing channel scheme is proposed for APT and YUM with the S3-hosted option:
+
+APT: `deb https://apt.<domain>/ <stable/testing/nightly> <v6/v7/v8/...>`
+
+YUM: `https://yum.<domain>/<stable/testing/nightly>/<v6/v7/v8/...>/`
+
+#### Backwards compatiblity
+To maintain backwards compatibility with our current solution we can reroute requests for the current repos to the new ones. This can be accomplished by placing a reverse proxy in front of the current repos and then using it to rewrite or redirect the incoming request to the new repos.
+
+### Future work
+While a specific solution is outside of the scope of this RFD, it is pertinent to discuss a disaster scenario that is common to all solutions, including the current one. If the hosting solution that contains the repo (i.e. a S3 bucket or GCP Artifact Registry) is deleted then all artifacts must be rebuilt and published from scratch. It looks like the Drone pipeline for Teleport takes around 90 minutes to run. Depending on how many instances can be ran at once without conflicting with each other, it could take several hours to get the repository back online to it's previous state. This could be alleviated by backing up artifacts after they're built, or by backing up the entire hosting solution. 
```

Comment:
I've got some ideas here.  DR backups are defiantly worth doing, with data replicated to a seperate region and versioning turned on.

Furthermore, I'd like to push towards a two phase publishing process:

1) upon tag (or maybe even push to `master`/`releases/*`) push the bits to a internal, private, staging repo.
2) upon promotion, re-sign with production keys, and push to the customer facing repo

Mostly, I want this so we can test most of the publishing without impacting customers.  This has been a pain point for me, and many others running releases, as discussed recently at https://github.com/gravitational/teleport/pull/10804#pullrequestreview-899453075